openedx_tagging/core/tagging/models/base.py

```diff
@@ -342,6 +342,70 @@ def get_filtered_tags(
 
         return tag_set.order_by("value", "id")
 
+    def add_tag(
+        self,
+        tag_value: str,
+        parent_tag_id: int | None = None,
+        external_id: str | None = None
+
+    ) -> Tag:
+        """
+        Add new Tag to Taxonomy. If an existing Tag with the `tag_value` already
+        exists in the Taxonomy, an exception is raised, otherwise the newly
+        created Tag is returned
+        """
+        self.check_casted()
+
+        if self.allow_free_text:
+            raise ValueError(
+                "add_tag() doesn't work for free text taxonomies. They don't use Tag instances."
+            )
+
+        if self.system_defined:
+            raise ValueError(
+                "add_tag() doesn't work for system defined taxonomies. They cannot be modified."
+            )
+
+        if self.tag_set.filter(value__iexact=tag_value).exists():
+            raise ValueError(f"Tag with value '{tag_value}' already exists for taxonomy.")
+
+        parent = None
+        if parent_tag_id:
+            parent = Tag.objects.get(id=parent_tag_id)
```

Comment:
We should ensure that the `parent` tag is in this taxonomy, e.g.

```suggestion
        if parent_tag_id:
            parent = self.tag_set.get(id=parent_tag_id)
```

Reply:
Also, we should also catch the `Tag.DoesNotExist` that this might throw and ensure the explanation gets back to the user [when calling the view](https://github.com/openedx/openedx-learning/pull/96/files#diff-215d20c663dcd5727c14740a72d09fe6fd22e3a92fd08db46036e2e799699917R549).

Reply:
Thanks for catching this! I updated it and added a test for this case, I initially was only testing for "invalid" tag ids forgot to check for the case of "valid but for other taxonomy".